### PR TITLE
Meters support in switch validation.

### LINKED ISSUE #1551: Meters support in switch validation
Current implementation
===================
Currently we have `/{switch_id}/rules/validate` API which validates only rules on the switch.

What needs to be done
==================
We also interested in checking meters on the switch.

Proposal
=======
Add new API path `/{switch_id}/validate`. That call will validate rules and meters and make the following response:
```
{
  "rules": {
    "missing": [<cookie>],
    "excess": [<cookie>],
    "proper": [<cookie>]
  },
  "meters": {
    "missing": [ {
        "meterId": 28,
        "cookie": <cookie>,
        "flowId": "foo"
      }
    ],
    "misconfigured": [ {
        "meterId": 32,
        "cookie": <cookie>,
        "flowId": "foo",
        "actualRate": 100,
        "expectedRate": 200
      }
    ],
    "excess": [ {
        "meterId": 15
      }
    ],
    "proper": [ {
        "meterId": 30,
        "cookie": <cookie>,
        "flowId": "foo"
      }
    ]
  }
}
```

Subtask of #1249

### PR DESCRIPTION
**Pull request by Settis**
Add new NB API point /{switch}/validate for validate
rules and meters on the switch.

Close #1551.

### CODE REVIEW

**Comment by rtretyak** [comment]
1. Do you take OF_12 switches into consideration? AFAIK we are unable to dump meters from them
2. What is the difference between misconfigured, missing and excess rule/meter?
3. Why are you putting *cookie* values into *meters*? I thought meters section will reported with proper/missing *meterIds*, not cookies

File: services/src/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java [hunk 1]
```diff
@@ -38,6 +41,32 @@ default RulesSyncResult toRulesSyncResult(RulesValidationResult validationResult
...
     RulesValidationResult toRulesValidationResult(SyncRulesResponse response);
 
+    /**
+     * Maps {@code}SyncRulesResponse{@code} to {@code}ValidationResult{@code}.
+     *
+     * @param response the {@code}SyncRulesResponse{@code}
+     * @return the result
+     */
+    default ValidationResult toValidationResult(SyncRulesResponse response) {
+        if (response == null) {
```
[the remark this answers is not in the document]
Comment (Settis) [reply services/src/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java]: It was taken from `SwitchMapperImpl`. I think it's more safety. From the other side it just postpone NPE.
OK, I'll delete it.

File: services/topology-engine/queue-engine/topologylistener/flow_utils.py [hunk 2]
```diff
@@ -515,40 +515,56 @@ def validate_switch_rules(switch_id, switch_rules):
     Perform validation of provided rules against the switch flows.
     """
 
-    switch_cookies = [x['cookie'] for x in switch_rules]
+    switch_cookies_with_meters = {x['cookie']: x['instructions']['instruction_goto_meter'] for x in switch_rules}
```
[the remark this answers is not in the document]
Comment (Settis) [reply services/topology-engine/queue-engine/topologylistener/flow_utils.py]: As I see from FL response, rules always have ['instructions']['instruction_goto_meter'] field. I'll fix it.
You are right, meters_by_cookie more clear.

**Comment by Settis** [comment]
@rtretyak, here is the answers:
> 1. Do you take OF_12 switches into consideration? AFAIK we are unable to dump meters from them

No, I don't know about it. Are you suggesting to skip meters check in this case? Do you think that we need additional flag in response which is show special case with no check?

> 2. What is the difference between misconfigured and missing + misconfigured and excess rule/meter? I don't really get what 'misconfigured' means. Are we checking the rule match itself to say that it's misconfigured? Are we checking the meter value itself to say that it's misconfigured?

I assume that a meter related to a rule. So, no meter can exists without a rule. In switch validation we not check rule's content, so 'misconfigured' meaningless for rules. Based on that I decided the following:
**Case 1** Flow from DB contains Cookie with Meter, but switch has Cookie without Meter. Then response will contain {proper_rules:{COOKIE}, misconfigured_meters:{COOKIE}}.
**Case 2** Flow from DB contains Cookie with Meter, but switch has nothing. Then response will contain {missing_rules:{COOKIE}, missing_meters:{COOKIE}}.
**Case 3** Flow segment from DB has Cookie, but switch has Cookie and Meter. Then response will contain {proper_rules:{COOKIE}, excess_meters:{COOKIE}}.
**Case 4** In DB we has no records, but switch has Cookie and Meter. Then response will contain {excess_rules:{COOKIE}, excess_meters:{COOKIE}}.

> 3. Why are you putting _cookie_ values into _meters_? I thought meters section will report with proper/missing _meterIds_, not cookies

I thought that meterId has no relation to Flow and using cookie will be clear here.

**Comment by Settis** [comment]
Current implementation works only with rules dump, but meters dump is needed also. So, whole logic must be written other way.

**Pull request closed without merging**